sentinel-core/src/main/java/com/alibaba/csp/sentinel/spi/SpiLoader.java

```diff
@@ -0,0 +1,525 @@
+                    }
+
+                    RecordLog.info("[SpiLoader]Found SPI,Service={},Provider={},aliasname={},isSingleton={},isDefault={},order={}",
+                        service.getName(), line, aliasName
+                            , spi == null ? true : spi.isSingleton()
+                            , spi == null ? false : spi.isDefault()
+                            , spi == null ? 0 : spi.order());
+                }
+
+                sortedClassList.addAll(classList);
+                Collections.sort(sortedClassList, new Comparator<Class<? extends S>>() {
+                    @Override
+                    public int compare(Class<? extends S> o1, Class<? extends S> o2) {
+                        Spi spi1 = o1.getAnnotation(Spi.class);
+                        int order1 = spi1 == null ? 0 : spi1.order();
+
+                        Spi spi2 = o2.getAnnotation(Spi.class);
+                        int order2 = spi2 == null ? 0 : spi2.order();
+
+                        return Integer.compare(order1, order2);
+                    }
+                });
+            } catch (IOException e) {
+                fail("error reading SPI configuration file", e);
+            } finally {
+                closeResources(in, br);
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "com.alibaba.csp.sentinel.spi.SpiLoader[" + service.getName() + "]";
+    }
+
+    /**
+     * Create Provider instance list
+     *
+     * @param clazzList class types of Providers
+     * @return Provider instance list
+     */
+    private List<S> createInstanceList(List<Class<? extends S>> clazzList) {
+        if (clazzList == null || clazzList.size() == 0) {
+            return Collections.emptyList();
+        }
+
+        List<S> instances = new ArrayList<>(clazzList.size());
+        for (Class<? extends S> clazz : clazzList) {
+            S instance = createInstance(clazz);
+            instances.add(instance);
+        }
+        return instances;
+    }
+
+    /**
+     * Create Provider instance
+     *
+     * @param clazz class type of Provider
+     * @return Provider class
+     */
+    private S createInstance(Class<? extends S> clazz) {
+        Spi spi = clazz.getAnnotation(Spi.class);
+        boolean singleton = true;
+        if (spi != null) {
+            singleton = spi.isSingleton();
+        }
+        return createInstance(clazz, singleton);
+    }
+
+    /**
+     * Create Provider instance
+     *
+     * @param clazz     class type of Provider
+     * @param singleton if instance is singleton or prototype
+     * @return Provider instance
+     */
+    private S createInstance(Class<? extends S> clazz, boolean singleton) {
+        S instance = null;
+        try {
+            if (singleton) {
+                instance = singletonMap.get(clazz.getName());
+                if (instance == null) {
+                    synchronized (this) {
```

Comment:
double check `instance = map.get(...)` may necessary, avoid `class.newInstance()` twice or more.

Reply:
Yes, thanks for pointing it out!

Reply:
It has been fixed, please check.

Reply:
After `synchronized (this)`, you should check `instance` not null again. same like this.
```java
instance = singletonMap.get(clazz.getName());
if (instance == null) {
    synchronized (this) {
        instance = singletonMap.get(clazz.getName());
        if (instance == null) {
            instance = service.cast(clazz.newInstance());
            singletonMap.put(clazz.getName(), instance);
        }
    }
}
```
may this implement correct, please take a look.